core/src/code.rs

```diff
@@ -91,3 +91,62 @@ impl CheckedCodeWithHash {
         (self.0, self.1)
     }
 }
+
+/// Contains instrumented binary code of a program and initial memory size from memory import.
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq)]
+pub struct InstrumentedCode {
+    code: Vec<u8>,
+    static_pages: u32,
+    instruction_weights_version: u32,
+}
+
+impl InstrumentedCode {
```

Comment:
I would like to have more type safety:

this struct should be built from `CheckedCode` and perform the code instrumentation. For example:
```rust
pub fn try_new(code: CheckedCode, version: u32, get_gas_rules: F) -> Result<Self>
where
     R: wasm_instrument::gas_metering::Rules,
     F: FnMut(&parity_wasm::elements::Module) -> R,
{
            let module = wasm_instrument::parity_wasm::deserialize_buffer(code.code())
                .map_err(|_| Error::FailedToDeserialize)?;
                
            let gas_rules = get_gas_rules(&module);

            let instrumented_module =
                wasm_instrument::gas_metering::inject(module, gas_rules, "env")
                    .map_err(|_| Error::FailedToInstrumentCode)?;

            let instrumented = wasm_instrument::parity_wasm::elements::serialize(instrumented_module)
                .map_err(|_| Error::FailedToSeriliazeModule)?;
                
            Self {
                checked_code: code,
                instrumented,
                version
            }
        }
        ```
        
The struct has to contain both the original checked code and the instrumented code. Also I suppose to store in the storage only `InstrumentedCode`. Implement `From<InstrumentedCode> for CheckedCode`.

`InstrumentedCodeWithHash` should be adjusted too: constructor should accept `InstrumentedCode` only and calculates hash for the internal checked code.